pypsa/statistics.py

```diff
@@ -386,20 +336,93 @@ def set_parameters(self, **kwargs: Any) -> None:
         """
         self.parameters.set_parameters(**kwargs)
 
+    @classmethod
+    def _get_grouping(
+        cls,
+        n: Network,
+        c: str,
+        groupby: Callable | Sequence[str] | str | bool,
+        port: str | None = None,
+        nice_names: bool = False,
+    ) -> dict:
+        by = None
+        level = None
+        if callable(groupby):
+            if "port" in signature(groupby).parameters:
+                by = groupby(n, c, port=port, nice_names=nice_names)
+            else:
+                by = groupby(n, c, nice_names=nice_names)
+        elif isinstance(groupby, list):
+            by = [n.static(c)[key] for key in groupby]
+        elif isinstance(groupby, str):
+            by = n.static(c)[groupby]
+        elif groupby is not False:
+            raise ValueError(
+                f"Argument `groupby` must be a function, list, string, False or dict, got {type(groupby)}"
+            )
+        return dict(by=by, level=level)
+
+    @property
+    def is_multi_indexed(self) -> bool:
+        return isinstance(self.n.snapshots, pd.MultiIndex)
+
+    @classmethod
+    def _aggregate_timeseries(
+        cls, obj: Any, weights: pd.Series, agg: str | Callable | bool = "sum"
+    ) -> Any:
+        """
+        Calculate the weighted sum or average of a DataFrame or Series.
+        """
+        if not agg:
+            return obj
+
+        if agg == "mean":
+            if isinstance(weights.index, pd.MultiIndex):
+                weights = weights.groupby(level=0).transform(lambda w: w / w.sum())
+            else:
+                weights = weights / weights.sum()
+            agg = "sum"
+
+        return cls._aggregate_with_weights(obj, weights, agg)
+
+    # The following methods are implemented in the concrete classes
+    @abstractmethod
+    def _aggregate_with_weights(self, *args: Any, **kwargs: Any) -> Any:
+        raise NotImplementedError
```

Comment:
```suggestion
    @abstractmethod
    def _aggregate_with_weights(self, *args: Any, **kwargs: Any) -> Any:
        pass
```
When using `@abstractmethod` just use pass, since abc is handeling the raised error already

Reply:
thanks, done